### PR TITLE
Nullsafety 

### PR DESCRIPTION
**Pull request by Schnurber**
Nullsafety with avoidance of nullable variables. All tests are working.

### CODE REVIEW

File: pubspec.yaml [hunk 1]
```diff
@@ -1,14 +1,19 @@
 name: tiled
-version: 0.6.0
+version: 0.8.0
```
Comment (Akida31) [inline pubspec.yaml:2]: why this bigger increase?

File: lib/src/flips.dart [hunk 2]
```diff
@@ -10,9 +8,9 @@ class Flips {
   const Flips.defaults() : this(false, false, false);
 
   Flips copyWith({
-    bool horizontally,
-    bool vertically,
-    bool diagonally,
+    bool? horizontally,
+    bool? vertically,
+    bool? diagonally,
```
Comment (spydon) [inline lib/src/flips.dart:13]: Couldn't these be non-nullable and have default values?

File: lib/src/layer.dart [hunk 3]
```diff
@@ -1,53 +1,54 @@
...
   static const int FLIPPED_DIAGONALLY_FLAG = 0x20000000;
 
-  String name;
-  int width;
-  int height;
-  bool visible;
+  String name = "";
+  int width = 0;
+  int height = 0;
+  bool visible = true;
```
Comment (spydon) [inline lib/src/layer.dart:17]: Why do these need initialization here when they also come in as non-nullable arguments in the constructor?

File: lib/src/layer.dart [hunk 4]
```diff
@@ -1,53 +1,54 @@
...
-
     NodeDSL.on(element, (dsl) {
       name = dsl.strOr('name', name);
       width = dsl.intOr('width', width);
       height = dsl.intOr('height', height);
       visible = dsl.boolOr('visible', true);
     });
 
-    final dataElement = element.children.firstWhere(
+    final XmlNode dataElement = element.children.firstWhere(
```
Comment (spydon) [inline lib/src/layer.dart:43]: No explicit type notations on local variables.

File: lib/src/layer.dart [hunk 5]
```diff
@@ -56,13 +57,14 @@ class Layer {
...
+    properties = TileMapParser.parsePropertiesFromElement(element);
   }
 
   // TMX data format documented here: https://github.com/bjorn/tiled/wiki/TMX-Map-Format#data
   void assembleTileMatrix(var bytes) {
-    tileMatrix = List.generate(height, (_) => List<int>(width));
-    tileFlips = List.generate(height, (_) => List<Flips>(width));
+    tileMatrix = List.generate(height, (_) => List<int>.filled(width, 0));
+    tileFlips = List.generate(height,
+        (_) => List<Flips>.filled(width, const Flips(false, false, false)));
```
Comment (spydon) [inline lib/src/layer.dart:67]: Not part of this PR, but these generate functions should be able to fill the lists with the correct values directly instead of having the loops afterwards.

File: lib/src/layer.dart [hunk 6]
```diff
@@ -1,53 +1,54 @@
...
-  List<List<int>> tileMatrix;
-  List<List<Flips>> tileFlips;
+  TileMap? map;
+  late List<List<int>> tileMatrix;
+  late List<List<Flips>> tileFlips;
 
-  List<List<Tile>> _tiles;
-  List<List<Tile>> get tiles {
+  List<List<Tile>>? _tiles;
+  List<List<Tile>>? get tiles {
```
Comment (spydon) [inline lib/src/layer.dart:26]: This shouldn't need the `?` right?

File: lib/src/object_group.dart [hunk 7]
```diff
@@ -1,32 +1,32 @@
...
+import 'tile_map_parser.dart';
+import 'tile_map.dart';
+import 'tmx_object.dart';
+import 'node_dsl.dart';
 
 class ObjectGroup {
-  String name;
-  String color;
+  String name= "";
+  String color = "";
```
Comment (spydon) [inline lib/src/object_group.dart:9]: What does it mean for these to default to empty strings?

File: lib/src/object_group.dart [hunk 8]
```diff
@@ -1,32 +1,32 @@
...
       color = dsl.strOr('color', color);
       opacity = dsl.doubleOr('opacity', opacity);
       visible = dsl.boolOr('visible', visible);
     });
+    properties = TileMapParser.parsePropertiesFromElement(element);
 
-    properties = TileMapParser._parsePropertiesFromElement(element);
-
-    final objectNodes = element.children
+    final Iterable<XmlElement> objectNodes = element.children
```
Comment (spydon) [inline lib/src/object_group.dart:29]: No types on local variables.

File: lib/src/tile.dart [hunk 9]
```diff
@@ -1,34 +1,38 @@
...
+  int gid = 0;
 
-  int width;
-  int height;
-  int spacing;
-  int margin;
+  int width = 0;
+  int height = 0;
+  int spacing = 0;
+  int margin = 0;
```
Comment (spydon) [inline lib/src/tile.dart:18]: These should be late right?

**Review by spydon — COMMENTED** [review]
Only commented 5 files so far, but I think many of the comments apply to patterns on the other files too. You shouldn't add values to fields that are later initialized in the constructor (add late/final late on them where it makes sense) and don't add explicit types on local variables.

File: pubspec.yaml (hunk 1 above)
Comment (Akida31) [reply pubspec.yaml:2]: and probably update the README according to the version

File: lib/src/layer.dart (hunk 3 above)
Comment (Schnurber) [reply lib/src/layer.dart:17]: There are two more constructors (onne named constructor ), and then code would be double, I think this is very clear to read.

File: lib/src/layer.dart (hunk 4 above)
Comment (Schnurber) [reply lib/src/layer.dart:43]: I agree!

File: lib/src/layer.dart (hunk 6 above)
Comment (Schnurber) [reply lib/src/layer.dart:26]: In this case I think it is a complex data structure, an it comes form somwhere and there is no meaningful simple initial value.

File: lib/src/object_group.dart (hunk 7 above)
Comment (Schnurber) [reply lib/src/object_group.dart:9]: I don't know, null is not much better. ;-)

File: lib/src/object_group.dart (hunk 8 above)
Comment (Schnurber) [reply lib/src/object_group.dart:29]: I agree!

File: lib/src/tile.dart (hunk 9 above)
Comment (Schnurber) [reply lib/src/tile.dart:18]: Maybe, but there is also `Tile.emptyTile()`, so it is easier, I think.

File: lib/src/flips.dart (hunk 2 above)
Comment (Schnurber) [reply lib/src/flips.dart:13]: I'm not sure, but bool types could then have three values, true, false, and null.

File: pubspec.yaml (hunk 1 above)
Comment (Schnurber) [reply pubspec.yaml:2]: Sorry...

File: lib/src/layer.dart (hunk 6 above)
Comment (spydon) [reply lib/src/layer.dart:26]: On this getter we know that it is initialized, so you can just return `_tiles!`

Comment (Schnurber) [reply lib/src/layer.dart:26]: Yes, the `?` could be removed for the getter, but what you want to return instead if `_tiles` is `null`?

Comment (spydon) [reply lib/src/layer.dart:26]: It can't be null in the getter, since it does `_recalculateTiles()` if it is null.

Comment (Schnurber) [reply lib/src/layer.dart:26]: ...it should be `_tiles!`

File: lib/src/tile.dart (hunk 9 above)
Comment (spydon) [reply lib/src/tile.dart:18]: Then the default values for an empty tile should be in the constructor.

**Comment by luanpotter** [comment]
Let me take a look at this, I will try to update it with current master
